doc/monads/functors.lean

```diff
@@ -0,0 +1,215 @@
+/-!
+
+In this `Functor`, `map` is called a "higher order function" because it takes a function as input
+`(α → β)` and returns another function as output `F α → F β`.
+
+Note that Lean is very powerful and lets you define your own syntax, so `<$>` is nothing special.
+You can define your own infix operator like this:
+
+-/
+infixr:100 " doodle " => Functor.map
+
+#eval (· * 5) doodle [1, 2, 3]  -- [5, 10, 15]
+/-!
+
+Let's look at some more examples:
+
+-/
+-- List String → List Nat
+#eval ["elephant", "tiger", "giraffe"].map (fun s => s.length)
+-- [8, 5, 7]
+
+-- List Nat → List Float
+#eval [1,2,3,4,5].map (fun s => (s.toFloat) ^ 3.0)
+-- [1.000000, 8.000000, 27.000000, 64.000000, 125.000000]
+
+--- List String → List String
+#eval (fun s => s.capitalize) <$> ["chris", "david", "mark"]
+-- ["Chris", "David", "Mark"]
+/-!
+
+Another example of a functor is the `Option` type. Option contains a value or nothing and is handy
+for code that has to deal with optional values, like optional command line arguments.
+
+Remember you can construct an Option using the type constructors `some` or `none`:
+
+-/
+#check some 5 -- Option Nat
+#eval some 5  -- some 5
+#eval (some 5).map (fun x => x + 1) -- some 6
+#eval (some 5).map (fun x => toString x) -- some "5"
+/-!
+
+Lean also provides a convenient short hand syntax for `(fun x => x + 1)`, namely `(· * 5)`
+using the middle dot unicode character which you can type in VS code using `\. `.
+
+-/
+#eval (some 4).map (· * 5)  -- some 20
+/-!
+
+The `map` function preserves the `none` state of the Option, so again
+map preserves the structure of the object.
+
+-/
+def x : Option Nat := none
+#eval x.map (fun x => toString x) -- none
+#check x.map (fun x => toString x) -- Option String
+/-!
+
+Notice that even in the `none` case it has transformed `Option Nat` into `Option String` as
+you see in the `#check` command.
+
+## How to make a Functor Instance?
+
+The `List` type is made a functor by the following `Functor` type class instance:
+
+-/
+instance : Functor List where
+  map := List.map
+/-!
+
+Notice all you need to do is provide the map function implementation.  For a quick
+example, let's supposed you create a new type describing the measurements of a home
+or apartment:
+
+-/
+structure LivingSpace (α: Type) where
+  totalSize : α
+  numBedrooms : Nat
+  masterBedroomSize : α
+  livingRoomSize : α
+  kitchenSize : α
+  deriving Repr, BEq
+/-!
+
+Now you can construct a `LivingSpace` in square feet using floating point values:
+-/
+abbrev SquareFeet := Float
+
+def mySpace : LivingSpace SquareFeet :=
+  { totalSize := 1800, numBedrooms := 4, masterBedroomSize := 500,
+    livingRoomSize := 900, kitchenSize := 400 }
+/-!
+
+Now, suppose you want anyone to be able to map a `LivingSpace` from one type of measurement unit to
+another.  Then you would provide a `Functor` instance as follows:
+
+-/
+def LivingSpace.map (f : α → β) (s : LivingSpace α) : LivingSpace β :=
+  { totalSize := f s.totalSize,
```

Comment:
We can avoid the `,`s.

Reply:
fixed in https://github.com/leanprover/lean4/pull/1505/commits/e53237c5a9afcfcfd2536283350598f393081ab3